File: app-db-master-slave/src/main/java/com/ccx/demo/business/user/vo/Authority.java
package com.ccx.demo.business.user.vo;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.annotation.JSONType;
import com.ccx.demo.business.user.enums.AuthorityCode;
import com.ccx.demo.business.user.enums.AuthorityType;
import com.support.mvc.entity.base.Prop;
import lombok.Data;
import lombok.SneakyThrows;
import lombok.experimental.Accessors;

import java.io.Serializable;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.support.mvc.entity.base.Prop.Type.*;

/**
 * <pre>
 * 菜单及操作权限
 *
 * @author 谢长春 2019/8/27
 */
@Data
@Accessors(chain = true)
@JSONType(orders = {"code", "parentCode", "name", "icon", "type", "url", "nodes"})
public class Authority implements Cloneable, Serializable {
    /**
     * Controller 方法头上的注解权限代码
     */
    private AuthorityCode code;
    /**
     * 父级代码
     */
    private AuthorityCode parentCode;
    /**
     * 名称
     */
    private String name;
    /**
     * 图标
     */
    private String icon;
    /**
     * 权限类型
     */
    private AuthorityType type;
    /**
     * 菜单路由地址 route，可选参数，需要跟前端约定
     */
    private String route;
    /**
     * 权限是否被选中
     */
    private boolean checked;
    /**
     * 子节点
     */
    private List<Authority> nodes;

    @Override
    public String toString() {
        return JSON.toJSONString(this);
    }

    @SneakyThrows
    @Override
    public Authority clone() {
        return (Authority) super.clone();
    }

    /**
     * 实体类所有属性名
     * 当其他地方有用到字符串引用该类属性时，应该使用该枚举定义
     */
    public enum Props {
        code(STRING.build("Controller 方法头上的注解权限代码")),
        parentCode(STRING.build("父级代码")),
        name(STRING.build("名称")),
        icon(STRING.build("图标")),
        type(ENUM.build("权限类型")),
        route(STRING.build("菜单路由地址 route，可选参数，需要跟前端约定")),
        checked(STRING.build("权限是否被选中")),
        nodes(ARRAY.build("子节点")),
        ;
        private final Prop prop;

        public Prop getProp() {
            return prop;
        }

        Props(final Prop prop) {
            prop.setName(this.name());
            this.prop = prop;
        }

        public static List<Prop> list() {
            return Stream.of(Props.values()).map(Props::getProp).collect(Collectors.toList());
        }
    }
}

File: gulp/src/api/entity/Authority.js
import AuthorityService from '../AuthorityService';

/**
 * 后台服务请求：权限指令
 * @author 谢长春 2019-7-28
 */
export default class Authority {
  /**
   * js 中， 类对象在经过方法传递后无法推断类型，造成类方法和变量提示不准确，这里 self 转换之后可以得到正确的提示
   * @param self {Authority}
   * @return {Authority}
   */
  static self(self) {
    return self;
  }

  /**
   * 构造函数：内部应该列出所有可能的参数，并对参数做说明
   * @param code {string} 权限指令代码
   * @param parentCode {string} 父级权限指令代码
   * @param name {string} 名称
   * @param icon {string} 图标
   * @param type {string} 权限类型
   * @param route {string} 菜单路由地址 route，可选参数，需要跟前端约定
   * @param checked {boolean} 权限是否被选中
   * @param nodes {Array<Authority>} 权限树子节点
   */
  constructor({
                code = undefined,
                parentCode = undefined,
                name = undefined,
                icon = undefined,
                type = undefined,
                route = undefined,
                checked = undefined,
                nodes = undefined
              } = {}) {
    /**
     * 权限指令代码
     * @type {string}
     */
    this.code = code;
    /**
     * 父级权限指令代码
     * @type {string}
     */
    this.parentCode = parentCode;
    /**
     * 名称
     * @type {string}
     */
    this.name = name;
    /**
     * 图标
     * @type {string}
     */
    this.icon = icon;
    /**
     * 权限类型
     * @type {string}
     */
    this.type = type;
    /**
     * 菜单路由地址 route，可选参数，需要跟前端约定
     * @type {string}
     */
    this.route = route;
    /**
     * 权限是否被选中
     * @type {boolean}
     */
    this.checked = checked;
    /**
     * 权限树子节点
     * @type {Array<Authority>}
     */
    this.nodes = nodes;
  }

  toString() {
    return JSON.stringify(this)
  }

  /**
   * 获取 api 服务对象
   * @return {AuthorityService}
   */
  getService() {
    return new AuthorityService(this);
  }
}
